read-fonts/src/tables/postscript/charstring.rs

```diff
@@ -1,203 +1,644 @@
+                            i += 1;
+                        }
+                        while i < self.stack.len() {
+                            let x1 = x;
+                            let y1 = y + self.stack.get_fixed(i)?;
+                            let x2 = x1 + self.stack.get_fixed(i + 1)?;
+                            let y2 = y1 + self.stack.get_fixed(i + 2)?;
+                            x = x2;
+                            y = y2 + self.stack.get_fixed(i + 3)?;
+                            sink.curve_to(x1, y1, x2, y2, x, y);
+                            i += 4;
+                        }
+                        self.stack.clear();
+                    }
+                    // Emits curves that start and end horizontal, unless
+                    // the stack count is odd, in which case the first
+                    // curve may start with a vertical tangent
+                    // <https://gitlab.freedesktop.org/freetype/freetype/-/blob/80a507a6b8e3d2906ad2c8ba69329bd2fb2a85ef/src/psaux/psintrp.c#L2789>
+                    HHCURVETO => {
+                        let mut i = 0;
+                        if self.stack.len_is_odd() {
+                            y += self.stack.get_fixed(0)?;
+                            i += 1;
+                        }
+                        while i < self.stack.len() {
+                            let x1 = x + self.stack.get_fixed(i)?;
+                            let y1 = y;
+                            let x2 = x1 + self.stack.get_fixed(i + 1)?;
+                            let y2 = y1 + self.stack.get_fixed(i + 2)?;
+                            x = x2 + self.stack.get_fixed(i + 3)?;
+                            y = y2;
+                            sink.curve_to(x1, y1, x2, y2, x, y);
+                            i += 4;
+                        }
+                        self.stack.clear();
+                    }
+                    // Alternates between curves with horizontal and vertical
+                    // tangents
+                    // <https://gitlab.freedesktop.org/freetype/freetype/-/blob/80a507a6b8e3d2906ad2c8ba69329bd2fb2a85ef/src/psaux/psintrp.c#L2834>
+                    HVCURVETO | VHCURVETO => {
+                        let count = self.stack.len();
+                        let mut i = (count & !2) - count;
+                        let mut alternate = op == HVCURVETO;
+                        while i < count {
+                            let (x1, x2, x3, y1, y2, y3);
+                            if alternate {
+                                x1 = x + self.stack.get_fixed(i)?;
+                                y1 = y;
+                                x2 = x1 + self.stack.get_fixed(i + 1)?;
+                                y2 = y1 + self.stack.get_fixed(i + 2)?;
+                                y3 = y2 + self.stack.get_fixed(i + 3)?;
+                                x3 = if count - i == 5 {
+                                    x2 + self.stack.get_fixed(i + 4)?
```

Comment:
Oops, this is missing an `i += 1`. Same for the similar branch below. I have a patch ready for this.

Reply:
Plausible to have tests that catch?

Reply:
Yes, I can add units tests for each path construction operator. Updated #478 to capture this.

Reply:
Added a test with a charstring that contains all of the path construction operators. We currently match FreeType. This will hopefully ensure we continue to do so.